src/mlpack/core/optimizers/katyusha/katyusha_impl.hpp

```diff
@@ -0,0 +1,202 @@
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+      std::numeric_limits<size_t>::max() : maxIterations;
+  for (size_t i = 0; i < actualMaxIterations; ++i)
+  {
+    // Calculate the objective function.
+    overallObjective = 0;
+    for (size_t f = 0; f < numFunctions; f += batchSize)
+    {
+      const size_t effectiveBatchSize = std::min(batchSize, numFunctions - f);
+      overallObjective += function.Evaluate(iterate0, f, effectiveBatchSize);
+    }
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "Katyusha: converged to " << overallObjective
+          << "; terminating  with failure.  Try a smaller step size?"
+          << std::endl;
+      return overallObjective;
+    }
+
+    if (std::abs(lastObjective - overallObjective) < tolerance)
+    {
+      Log::Info << "Katyusha: minimized within tolerance " << tolerance
+          << "; terminating optimization." << std::endl;
+      return overallObjective;
+    }
+
+    lastObjective = overallObjective;
+
+    // Compute the full gradient.
+    size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+    function.Gradient(iterate, 0, fullGradient, effectiveBatchSize);
+    for (size_t f = effectiveBatchSize; f < numFunctions;
+        /* incrementing done manually */)
+    {
+      // Find the effective batch size (the last batch may be smaller).
+      effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+      function.Gradient(iterate0, f, gradient, effectiveBatchSize);
+      fullGradient += gradient;
+
+      f += effectiveBatchSize;
+    }
+    fullGradient /= (double) numFunctions;
+
+    // To keep track of where we are and how things are going.
+    double cw = 1;
+    w.zeros();
+
+    for (size_t f = 0, currentFunction = 0; f < innerIterations;
+        /* incrementing done manually */)
+    {
+      // Is this iteration the start of a sequence?
+      if ((currentFunction % numFunctions) == 0)
+      {
+        currentFunction = 0;
+
+        // Determine order of visitation.
+        if (shuffle)
+          function.Shuffle();
+      }
+
+      // Find the effective batch size (the last batch may be smaller).
+      effectiveBatchSize = std::min(batchSize, numFunctions - currentFunction);
+      iterate = tau1 * z + tau2 * iterate0 + (1 - tau1 - tau2) * y;
+
+      // Calculate variance reduced gradient.
+      function.Gradient(iterate, currentFunction, gradient,
+          effectiveBatchSize);
+      function.Gradient(iterate0, currentFunction, gradient0,
+          effectiveBatchSize);
+
+      arma::mat zNew = z - alpha * (fullGradient + (gradient - gradient0) /
+          (double) batchSize);
+
+      // Proximal update, choose between Option I and Option II. Shift relative
+      // to the Lipschitz constant or take a constant step using the given step
+      // size.
+      if (proximal)
+      {
+        y = iterate + 1.0 / (3.0 * lipschitz) * (zNew - z);
```

Comment:
I guess there is a conversion in the paper too for line 11 to this, but I don't seem to be able to find it (I have also not read the paper in full detail, mostly just read the algorithm and skipped the proofs); do you want to add a note about it?